Add draft of DevOps Notification post (#1496)
Azure/azure-sdk: _posts/2020-06-30-devops-notifications.md

---
title: Routing Failure Notifications in DevOps
layout: post
date: 2020-06-30
sidebar: releases_sidebar
author_github: danieljurek
repository: azure/azure-sdk
---

> The Azure SDK Engineering System runs roughly 480 Azure DevOps pipelines to validate GitHub Pull Requests, run live test automation, and ensure that the libraries are ready to be released. If a build or test fails, our alerting system immediately notifies our SDK engineering teams and partner Azure service teams. In this post, we'll learn how The Azure SDK Engineering Systems team scaled Azure DevOps notifications to alert the right people at the right time when our automation discovers an issue.

Reviewer: ```suggestion
The Azure SDK Engineering System runs roughly 480 Azure DevOps pipelines to validate GitHub Pull Requests, run live test automation, and ensure that the libraries are ready to be released. If a build or test fails, our alerting system immediately notifies our SDK engineering teams and partner Azure service teams. In this post, we'll learn how the Azure SDK Engineering Systems team scaled Azure DevOps notifications to alert the right people at the right time when our automation discovers an issue.
```

## Alert System Architecture

The Engineering Systems team looks for existing tools to handle problems and we discovered that the Azure DevOps alerting system does most of the work for us. Rather than build our own alerting system we automate configuring Azure DevOps notification groups.

Our alerting system makes use of GitHub’s CODEOWNERS file that teams are already using to add reviewers to pull requests. These same owners are responsible for fixing code in their service’s SDK when a build or test fails. In order to route failure notifications we need to do a few things:

1. Get a list of pipelines with schedule triggers
1. Create notification groups for each pipeline and subscribe those groups to receive failure notifications
1. Synchronize appropriate code owners into the group

### Get a list of pipelines with schedule triggers

> This is a straightforward process of filtering pipelines in the Azure DevOps build definition API.

Reviewer: Maybe a code snippet or a little further explanation here would be good.

### Create notification groups

We query groups in DevOps and use YAML in the group description to understand each group’s purpose. This keeps us from having to store data in another database. The syntax is simple:

```yaml
pipelineId: 123
purpose: ParentNotificationTeam
```

The `pipelineId` field gives the Azure DevOps pipeline ID and the `purpose` can either be `ParentNotificationTeam` or `SynchronizedNotificationTeam`.

> We use two teams because some interested parties (like project leads or Engineering Systems administrators) may want to know when pipeline executions fail. The Parent Notification Team allows us to manually add these parties. The Synchronized Notification team gets the code owners.

Reviewer: I'm not sure what you are trying to say with this:  The Synchronized Notification team gets the code owners.

## Other things to try

Some ideas we had when designing this project but have not implemented yet:

**Microsoft Teams** – The Azure SDK team makes considerable use of Microsoft Teams and putting these alerts in a Teams channel could work better than email for the ways in which some teams work.

**Making better use of error logs** – Azure DevOps has a Logging Commands feature where a specifically formatted message sent to stdout will generate an alert in Azure DevOps. These alerts are the only detail a developer will see in an email and a little more effort here can help a product engineer quickly make sense of the failure and save time on investigations:

![An example of error message logs in an Azure DevOps failure email](https://devblogs.microsoft.com/azure-sdk/wp-content/uploads/sites/58/2020/06/devops-email-error-messages.png)

> ## Conclusion

Reviewer: See footer on this post.  Probably need @adrianhall to comment if he wants that as the last Dep Inj one did not have it.

Author: Thanks! I've added the footer from that release. I'll work with someone on the PM team to finalize that wording.

> Using and enforcing consistent pipeline and repo layouts means that our Engineering System will scale smoothly to meet the needs of our product engineers.  Small tools like this one can tune your system in ways that help your product teams stay focused on their goals.

Reviewer: Great.  There's also usually a blurb that is standard for all posts...like links and where to find more info.

---

> title: Routing failure notifications in DevOps

Reviewer: ```suggestion
title: Routing Failure Notifications in DevOps
```

layout: post
date: 2020-06-30
sidebar: releases_sidebar
author_github: danieljurek
repository: azure/azure-sdk
---

> Every day some 441 Azure DevOps pipelines validate GitHub Pull Requests, run live test automation, and ensure that code in our repositories’ main branches is ready for release. If a build or test fails, our alerting system lets our language engineering teams and partner service teams know right away. Here’s how we’ve scaled Azure DevOps notifications to alert the right people to an issue our automation discovers.

Author: The number 441 should be updated to a more recent figure before release

Reviewer 2: I wouldn't even given an exact number. I would say something like "every day about ~500 or so Azure DevOps pipelines".

> Every day some 441 Azure DevOps pipelines validate GitHub Pull Requests, run live test automation, and ensure that code in our repositories’ main branches is ready for release. If a build or test fails, our alerting system lets our language engineering teams and partner service teams know right away. Here’s how we’ve scaled Azure DevOps notifications to alert the right people to an issue our automation discovers.

Reviewer: ```suggestion
The Azure SDK engineering system runs roughly 450 Azure DevOps pipelines to validate GitHub Pull Requests, run live test automation, and ensure that the libraries are ready to be released. If a build or test fails, our alerting system immediately notifies our SDK engineering teams and partner Azure service teams. In this post, we'll learn how we've scaled Azure DevOps notifications to alert the right people at the right time when our automation discovers an issue.
```

## That’s a lot of pipelines

> We generally create one pipeline per service and language to build and test code relating to that pipeline. This keeps our DevOps agent machine time focused more narrowly and enables engineers to iterate more quickly on changes. There’s no need to build packages and run tests for Storage when you’re making changes to Key Vault.

Reviewer: ```suggestion
We generally create one pipeline per service/language combination to build and test code relating to that pipeline. This keeps our DevOps agent machine time focused more narrowly and enables engineers to iterate more quickly on changes. For example, there’s no need to build packages and run tests for Storage when you’re making changes to Key Vault.
```

> Keeping this many pipelines organized requires a standard approach to how we create, configure, name, and trigger pipelines. When you can apply standards you can automate work, so we built automation that using the [Azure DevOps API](https://docs.microsoft.com/en-us/rest/api/azure/devops/?view=azure-devops-rest-5.1). A small team of engineers can scale our engineering systems to meet the needs of an ever growing matrix of services and languages.

Reviewer: ```suggestion
Keeping this many pipelines organized requires a standard approach to how we create, configure, name, and trigger pipelines. When you can apply standards you can automate work, so we built automation that uses the [Azure DevOps API](https://docs.microsoft.com/rest/api/azure/devops/?view=azure-devops-rest-5.1). A small team of engineers can scale our engineering systems to meet the needs of an ever growing matrix of services and languages.
```

## Alert System Architecture

> The Engineering Systems team looks for existing tools to handle problems and, in this case, the Azure DevOps alerting system does most of the work for us. Rather than build our own alerting system we automate configuring Azure DevOps notification groups.

Reviewer: if you cap Eng Sys here, you should do it every where.

> The Engineering Systems team looks for existing tools to handle problems and, in this case, the Azure DevOps alerting system does most of the work for us. Rather than build our own alerting system we automate configuring Azure DevOps notification groups.

Reviewer: ```suggestion
The Engineering Systems team looks for existing tools to handle problems and we discovered that the Azure DevOps alerting system does most of the work for us. Rather than build our own alerting system we automate configuring Azure DevOps notification groups.
```

Our alerting system makes use of GitHub’s CODEOWNERS file that teams are already using to add reviewers to pull requests. These same owners are responsible for fixing code in their service’s SDK when a build or test fails. In order to route failure notifications we need to do a few things:

1. Get a list of pipelines with schedule triggers

> 1. Create notification groups for each pipeline we care about and subscribe those groups to receive failure notifications

Reviewer: ```suggestion
1. Create notification groups for each pipeline and subscribe those groups to receive failure notifications
```

Our pipeline definition yml files exist at the level of the service (e.g. [/sdk/appconfiguration/ci.yml](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/appconfiguration/ci.yml)) and matches the path leading up to the .yml file (e.g. `/sdk/appconfiguration/`) against the service team folder for each pipeline definition.

We translate the GitHub aliases using a Microsoft database of employees who contribute on behalf of the company and synchronize those contacts into the Synchronized Notification Team.

## Use our tools

The notification creation tool is open source so you can use it to build your own notification strategy. It is designed to work with repositories that follow our repo layout and pipeline structure but can be modified and extended to work with other environments. You can find the code at:

[https://github.com/Azure/azure-sdk-tools/tree/master/tools/notification-configuration](https://github.com/Azure/azure-sdk-tools/tree/master/tools/notification-configuration)

> If you are working outside of Microsoft you will need to adapt the [GitHubNameResolver](https://github.com/Azure/azure-sdk-tools/blob/master/tools/notification-configuration/github-codeowner-subscriber/GitHubNameResolver.cs) class to work with your own contact resolving strategy.

Reviewer 2: I think folks could simply send email to github users but the issue really boils down to mapping from github to devops but hopefully that issue will go away overtime.

## Other things to try
Some ideas we had when designing this project but have not implemented yet:

**Microsoft Teams** – The Azure SDK team makes considerable use of Microsoft Teams and putting these alerts in a Teams channel could work better than email for the ways in which some teams work.

> **Making better use of error logs** – Azure DevOps has a Logging Commands feature where a specifically formatted message sent to stdout will generate an alert in Azure DevOps. These alerts are the only detail a developer will see in an email and a little more effort here can help a product engineer quickly make sense of the failure and save time on investigations:

Reviewer: I recommend having a closing statement, summary, etc at the end.
